File: src-Kotlin/CycleSort.kt
class CycleSort {
    // Main function to test the Cycle sort function
    fun main(args: Array<String>) {
        val arr = intArrayOf(1, 8, 3, 9, 10, 10, 2, 4)
        val n = arr.size
        cycleSort(arr, n)

        System.out.println("After sort: ")
        for (i in 0 until n) {
            System.out.println(arr[i].toString() + " ")
        }
    }

    // Function sort the array using Cycle sort
    fun cycleSort(arr: IntArray, n: Int) {
        var w = 0 // count number of memory writes

        for (start in 0..n - 2) {
            var item = arr[start] // initializing the starting point

            var pos = start // find the position where we put the item
            for (i in start + 1 until n)
                if (arr[i] < item)
                    pos++

            if (pos == start)
            // if the item is already in the correct position
                continue

            while (item == arr[pos])
            // ignore all duplicate elements
                pos += 1

            if (pos != start) { // put the item to it's right position
                val temp = item
                item = arr[pos]
                arr[pos] = temp
                w++
            }

            while (pos != start) { // rotate rest of the cycle
                pos = start

                for (i in start + 1 until n)
                    if (arr[i] < item)
                        pos += 1

                while (item == arr[pos])
                    pos += 1

                if (item != arr[pos]) {
                    val temp = item
                    item = arr[pos]
                    arr[pos] = temp
                    w++
                }
            }
        }
    }
}

File: Python/CycleSort.py
# Python program to implement cycle sort
from typing import List, TypeVar

T = TypeVar('T')  # Create a generic type of T


def cycle_sort(array: List[T]) -> int:
    writes: int = 0

    # Loop through the array to find cycles to rotate.
    for cycleStart in range(0, len(array) - 1):
        item: T = array[cycleStart]

        # Find where to put the item.
        pos: int = cycleStart
        for i in range(cycleStart + 1, len(array)):
            if array[i] < item:
                pos += 1

        # If the item is already there, this is not a cycle.
        if pos == cycleStart:
            continue

        # Otherwise, put the item there or right after any duplicates.
        while item == array[pos]:
            pos += 1
        array[pos], item = item, array[pos]
        writes += 1

        # Rotate the rest of the cycle.
        while pos != cycleStart:

            # Find where to put the item.
            pos = cycleStart
            for i in range(cycleStart + 1, len(array)):
                if array[i] < item:
                    pos += 1

            # Put the item there or right after any duplicates.
            while item == array[pos]:
                pos += 1
            array[pos], item = item, array[pos]
            writes += 1

    return writes


# driver code
if __name__ == '__main__':
    arr = [1, 8, 3, 9, 10, 10, 2, 4]
    n = len(arr)
    cycle_sort(arr)

    print("After sort : ", arr)
